filestorage.go

```diff
@@ -149,33 +151,46 @@ func dataDir() string {
 	return filepath.Join(baseDir, "certmagic")
 }
 
-// TryLock attempts to get a lock for name, otherwise it returns
-// a Waiter value to wait until the other process is finished.
-func (fs FileStorage) TryLock(key string) (Waiter, error) {
+// Lock obtains a lock named by the given key. It blocks
+// until the lock can be obtained or an error is returned.
+func (fs FileStorage) Lock(key string) error {
+	// can't defer the unlock because we might have
+	// to Wait() for a while before returning, so we're
+	// careful to unlock at all the right places
 	fileStorageNameLocksMu.Lock()
-	defer fileStorageNameLocksMu.Unlock()
+
+	if fs.fileStorageNameLocks == nil {
+		fs.fileStorageNameLocks = make(map[string]*fileStorageWaiter)
```

Comment:
`fs` is a `copy` so this assignment has no external effect.

I fixed it in https://github.com/DisposaBoy/certmagic/commit/36c74006a73e70a707b23c35828dd146f3b3c4ea so I could continue testing.

Reply:
Dang, you're right. Any chance we can still create `FileStorage` without them having to be a pointer? I think the Go compiler takes care of passing in the reference in that case: https://tour.golang.org/methods/4 -- just one less thing to worry about then.

Reply:
That won't work: The method set of `FileStorage{}` only has methods with a `FileStorage` receiver so it doesn't implement the interface (`&FileStorage{}` has both `*FileStorage` and `FileStorage` methods).

Reply:
Alright. That's something I think I can live with. (using &FileStorage)